File: solutions/121.best-time-to-buy-and-sell-stock/best-time-to-buy-and-sell-stock.js
/**
 * @param {number[]} prices
 * @return {number}
 */
const maxProfit = function (prices) {

  let dp = [0] // profit
  let low = prices[0]

  for (let i = 1; i < prices.length; i++) {
    let price = prices[i]

    dp[i] = Math.max(dp[i - 1], price - low)
    low = Math.min(low, price)
  }

  return dp.pop()
}

File: solutions/121.best-time-to-buy-and-sell-stock/best-time-to-buy-and-sell-stock.py
class Solution:
    def maxProfit(self, prices):
        """
        :type prices: List[int]
        :rtype: int
        """
        if not prices:
            return 0

        low = prices[0]
        high = prices[0]

        ret = 0

        for p in prices[1:]:
            if p <= low:
                # find lower, cal cur profit
                low = p
                high = p
            else:
                # find higher, set new higher
                high = p
                ret = max(ret, high - low)

        return ret
